### PR TITLE
[Merged by Bors] - chore: use string gaps

### PR DESCRIPTION
**Pull request by kmill**
For error messages that span multiple lines, now we can use the new "string gap" escape sequence (a `\` at the end of the line, which causes all the whitespace at the beginning of the next line ignored), rather than using append operations or hacks involving `{` and `}` for `s!` and `m!` strings.

Style-wise, we suggest for long messages that the body of the message start on the next line if possible, by starting the message with a string gap sequence. This eliminates needing to decide whether to indent continuing lines at the `"` or at the next column. For example:
```lean
    logError m!"\
      Tactic 'more_magic' cannot \
      apply expression{indentD e}\n\
      due to some reason or another."
```
Note that to get whitespace to appear between one line and the next, you include it before the `\`.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by kmill** [comment]
@fpvandoorn With the newest Lean we don't need the `{""}` hack anymore :smile:

(I haven't labeled this `awaiting-review` yet because there's a PR I put on the queue that's going to immediately conflict with this one.)

File: Mathlib/Data/PNat/Xgcd.lean
Comment (digama0) [inline Mathlib/Data/PNat/Xgcd.lean:75]: the spaces look a bit inconsistent here (also in the original), is this deliberate?

File: Mathlib/Tactic/Eqns.lean
Comment (digama0) [inline Mathlib/Tactic/Eqns.lean:42]:
```suggestion
        throwError "There already exist stored eqns for '{declName}' registering new equations\n\
          will not have the desired effect."
```

File: Mathlib/Tactic/Simps/Basic.lean
Comment (digama0) [inline Mathlib/Tactic/Simps/Basic.lean:1050]:
```suggestion
        The definition {nm} is not a constructor application. Please use `@[simps!]` instead.\n\
        \n\
        Explanation: `@[simps]` uses the definition to find what the simp lemmas should \
```

File: Mathlib/Tactic/Simps/Basic.lean
Comment (digama0) [inline Mathlib/Tactic/Simps/Basic.lean:1113]:
```suggestion
    throwError "Invalid simp lemma {simpLemma}. \
      Structure {str} does not have projection {neededProj}.\n\
      The known projections are:\n  \
        {projs}\n\
      You can also see this information by running\n  \
        `initialize_simps_projections? {str}`.\n\
      Note: these projection names might be customly defined for `simps`, \
      and could differ from the projection names of the structure."
```

File: Mathlib/Tactic/ToAdditive.lean
Comment (digama0) [inline Mathlib/Tactic/ToAdditive.lean:621]: suggestion doesn't work, but consider hanging the `m!"\` here instead of `<|` on the previous line

File: Mathlib/Tactic/ToAdditive.lean
Comment (digama0) [inline Mathlib/Tactic/ToAdditive.lean:627]:
```suggestion
          declaration."
      else ""
```

File: Mathlib/Tactic/Variable.lean
Comment (digama0) [inline Mathlib/Tactic/Variable.lean:220]:
```suggestion
        bug, or you can try adjusting `set_option variable?.maxSteps {maxSteps}`\n\n\
        Current variable command:{indentD (← `(command| variable $binders'*))}"
```

File: Mathlib/Tactic/Widget/Calc.lean
Comment (digama0) [inline Mathlib/Tactic/Widget/Calc.lean:93]:
```suggestion
      s!"_ {relStr} {newLhsStr} := by sorry\n\
         {spc}_ {relStr} {newRhsStr} := by sorry\n\
         {spc}_ {relStr} {rhsStr} := by sorry"
```

File: test/GCongr/inequalities.lean
Comment (digama0) [inline test/GCongr/inequalities.lean:107]:
```suggestion
    "rel failed, cannot prove goal by 'substituting' the listed relationships. \
     The steps which could not be automatically justified were:\n0 ≤ x\nc ≤ d"
```

**Review by digama0 — APPROVED** [review]

File: Mathlib/Tactic/Eqns.lean
Comment (kmill) [reply Mathlib/Tactic/Eqns.lean:42]: I slipped in this message change. Having a newline there at that position didn't make sense to me.

@eric-wieser Do you want this newline?

File: Mathlib/Tactic/Simps/Basic.lean
Comment (kmill) [reply Mathlib/Tactic/Simps/Basic.lean:1113]: There needs to be a `toMessageData` in there to use `indentD`, so here's another suggestion to be able to add it to the batch. I also am not sure about stowing away indentation on the previous line, so here's an alternative.

```suggestion
    throwError "Invalid simp lemma {simpLemma}. \
      Structure {str} does not have projection {neededProj}.\n\
      The known projections are:\
      {indentD <| toMessageData projs}\n\
      You can also see this information by running\
      \n  `initialize_simps_projections? {str}`.\n\
      Note: these projection names might be customly defined for `simps`, \
      and could differ from the projection names of the structure."
```

File: Mathlib/Data/PNat/Xgcd.lean
Comment (kmill) [reply Mathlib/Data/PNat/Xgcd.lean:75]: I don't know what's going on with the additional spaces in the original. I got as far as figuring out that it's `[2x2 matrix, vector]` and tried lining it up in the source code.

File: Mathlib/Tactic/ToAdditive.lean
Comment (kmill) [reply Mathlib/Tactic/ToAdditive.lean:621]: I was thinking of hanging `m!"\`, but I thought that it would make it less clear that there's the `++ if` logic later. I think I'd rather leave this as a direct translation and then let Floris deal with this the next time he touches this part of ToAdditive.

File: Mathlib/Tactic/Eqns.lean
Comment (kmill) [reply Mathlib/Tactic/Eqns.lean:42]: @eric-wieser ping

Comment (eric-wieser) [reply Mathlib/Tactic/Eqns.lean:42]: No, I agree this newline should be removed

Comment (eric-wieser) [reply Mathlib/Tactic/Eqns.lean:42]: Though it looks like there is a mistake in this message:
```suggestion
        throwError "There already exist stored eqns for '{declName}'; registering new equations \
          will not have the desired effect."
```

File: Mathlib/Data/PNat/Xgcd.lean
Comment (kmill) [reply Mathlib/Data/PNat/Xgcd.lean:75]: @digama0 Should we try to fix the whitespace for this repr instance before merging? or could we leave it as-is?

Comment (digama0) [reply Mathlib/Data/PNat/Xgcd.lean:75]: I think unless more information is forthcoming it should be fine to just remove the spaces on the first line.

Comment (kmill) [reply Mathlib/Data/PNat/Xgcd.lean:75]: Done

Comment (kmill) [reply Mathlib/Data/PNat/Xgcd.lean:75]: @digama0 Do want to make any more changes, or are you happy with sending this PR onto the queue?

(Just pinging you because I'm not sure what your approval means precisely, whether that means you're waiting for another maintainer's opinion, whether it means you've delegated it to me, or whether it would be ok if I sent this to the maintainer merge queue myself.)

**Pull request closed without merging**
